src/Dictionaries/HashedDictionary.cpp

```diff
@@ -35,16 +41,151 @@ namespace ErrorCodes
+        for (size_t shard = 0; shard < shards; ++shard)
+            shards_blocks[shard] = block.cloneEmpty();
+
+        IColumn::Selector selector = createShardSelector(block, shards_slots);
+        splitBlock(selector, block, shards_blocks);
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            if (!shards_queues[shard]->push(std::move(shards_blocks[shard])))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to shards queue #{}", shard);
+        }
+    }
+
+    void finish()
+    {
+        for (auto & queue : shards_queues)
+            queue->finish();
+
+        Stopwatch watch;
+        pool.wait();
+        UInt64 elapsed_us = watch.elapsedMicroseconds();
+        LOG_TRACE(dictionary.log, "Processing the tail took {}ms", elapsed_us);
+    }
+
+    ~ParallelDictionaryLoader()
+    {
+        for (auto & queue : shards_queues)
+            queue->clearAndFinish();
+        pool.wait();
+    }
+
+private:
+    HashedDictionary & dictionary;
+    const size_t shards;
+    bool simple_key;
+    ThreadPool pool;
+    std::vector<std::optional<Queue>> shards_queues;
+    std::vector<UInt64> shards_slots;
+
+    void threadWorker(size_t shard)
+    {
+        Block block;
+        auto & shard_queue = *shards_queues[shard];
+
+        while (shard_queue.pop(block))
+        {
+            Stopwatch watch;
+            dictionary.blockToAttributes(block, shard);
+            UInt64 elapsed_us = watch.elapsedMicroseconds();
+            if (elapsed_us > 1'000'000)
+                LOG_TRACE(dictionary.log, "Block processing for shard #{} is slow {}ms (rows {}).", shard, elapsed_us, block.rows());
+        }
+
+        if (!shard_queue.isFinished())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not pull non finished shards queue #{}", shard);
+    }
+
+    /// Split block to shards smaller block, using 'selector'.
+    void splitBlock(const IColumn::Selector & selector, const Block & block, Blocks & splitted_blocks)
+    {
+        size_t columns = block.columns();
+        for (size_t col = 0; col < columns; ++col)
+        {
+            MutableColumns splitted_columns = block.getByPosition(col).column->scatter(shards, selector);
+            for (size_t shard = 0; shard < shards; ++shard)
+                splitted_blocks[shard].getByPosition(col).column = std::move(splitted_columns[shard]);
+        }
+    }
+
+    IColumn::Selector createShardSelector(const Block & block, const std::vector<UInt64> & slots)
+    {
+        size_t num_rows = block.rows();
+        IColumn::Selector selector(num_rows);
+
+        size_t skip_keys_size_offset = dictionary.dict_struct.getKeysSize();
+        Columns key_columns;
+        key_columns.reserve(skip_keys_size_offset);
+        for (size_t i = 0; i < skip_keys_size_offset; ++i)
+            key_columns.emplace_back(block.safeGetByPosition(i).column);
+
+        DictionaryKeysArenaHolder<dictionary_key_type> arena_holder;
```

Comment:
Arena probably can be reused if this function is called from single thread. Even if this is called from multiple threads, then we need to create array with arenas. Otherwise for complex keys there will be a lot of allocations.